File: backend/app/database/models/resource.py
from datetime import datetime

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Integer, String, Text, func, text
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class Resource(Base):
    __tablename__ = "resources"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    organization_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("organizations.id", ondelete="CASCADE"), index=True, nullable=False
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    storage_key: Mapped[str] = mapped_column(String(512), nullable=False)
    visibility: Mapped[str] = mapped_column(
        String(20), nullable=False, default="PRIVATE", server_default=text("'PRIVATE'")
    )
    created_by: Mapped[int | None] = mapped_column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"), index=True, nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    __table_args__ = (
        CheckConstraint("visibility IN ('PUBLIC', 'PRIVATE')", name="ck_resource_visibility"),
    )

File: frontend/src/types/resource.ts
export type ResourceVisibility = 'PUBLIC' | 'PRIVATE'

export interface Resource {
  id: number
  organization_id: number
  title: string
  description: string | null
  storage_key: string
  visibility: ResourceVisibility
  created_by: number | null
  created_at: string
  updated_at: string
}

export interface ResourceCreate {
  title: string
  description?: string
  storage_key: string
  visibility?: ResourceVisibility
}

export interface ResourceUpdate {
  title?: string
  description?: string | null
  storage_key?: string
  visibility?: ResourceVisibility
}
